controller-client/src/paginator.rs

```diff
@@ -0,0 +1,121 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+use crate::{ControllerClient, ControllerError, ResultRetry};
+use futures::prelude::*;
+use futures::stream::{self};
+use pravega_client_retry::retry_result::RetryError;
+use pravega_client_shared::Scope;
+use std::vec::IntoIter;
+use tracing::info;
+
+///
+///Helper method to iterated over the all the Pravega streams under the provided Scope.
+///This method returns a stream of values,Pravega streams, produced asynchronously.
+///
+/// The below snippets show case the example uses.
+/// Sample 1:
+///```ignore
+/// # futures::executor::block_on(async {
+/// use pravega_client_shared::Scope;
+/// use pravega_controller_client::paginator::list_streams;
+/// use pravega_client::client_factory::ClientFactory;
+/// use pravega_client_config::ClientConfigBuilder;
+/// use pravega_client_config::MOCK_CONTROLLER_URI;
+///     let config = ClientConfigBuilder::default()
+///         .controller_uri(MOCK_CONTROLLER_URI)
+///         .build()
+///         .expect("creating config");
+///     let controller_client = ClientFactory::new(config).get_controller_client();
+///     let stream = list_streams(
+///         Scope {
+///             name: "testScope".to_string(),
+///         },
+///         controller_client,
+///     );
+///     // collect all the Streams in a single vector
+///     let stream_list:Vec<String> = stream.map(|str| str.unwrap()).collect::<Vec<String>>().await;
+///  # });
+/// ```
+///
+/// Sample 2:
+/// ```ignore
+/// # futures::executor::block_on(async {
+/// use pravega_client_shared::Scope;
+/// use pravega_controller_client::paginator::list_streams;
+/// use pravega_client::client_factory::ClientFactory;
+/// use pravega_client_config::ClientConfigBuilder;
+/// use pravega_client_config::MOCK_CONTROLLER_URI;
+/// use futures::StreamExt;
+///     let config = ClientConfigBuilder::default()
+///         .controller_uri(MOCK_CONTROLLER_URI)
+///         .build()
+///         .expect("creating config");
+///     let controller_client = ClientFactory::new(config).get_controller_client();
+///     let mut stream = list_streams(
+///         Scope {
+///             name: "testScope".to_string(),
+///         },
+///         controller_client,
+///     );
+/// let pravega_stream_1 = stream.next().await;
+/// let pravega_stream_2 = stream.next().await;
+/// // A None is returned at the end of the stream.
+///  # });
+/// ```
+///
+pub fn list_streams(
+    scope: Scope,
+    client: &dyn ControllerClient,
+) -> impl Stream<Item = Result<String, RetryError<ControllerError>>> + '_ {
+    struct State {
+        streams: IntoIter<String>,
+        scope: Scope,
+        token: String,
+    }
+
+    // Initial state with an empty Continuation token.
+    stream::unfold(
+        State {
+            streams: Vec::new().into_iter(),
+            scope,
+            token: String::from(""),
+        },
+        move |mut state| async move {
```

Comment:
Make this a private function and pass the method here, rather than creating an anonymous one. It will look cleaner.

Reply:
improved it.